src/query/storage/cache/redis_cache.go

```diff
@@ -435,9 +502,40 @@ func BucketWindowGetOrFetch(
 			return storage.PromResult{}, nil
 		}
 		// Combine all the results together
-		return *combineResult(append(results, &res)), nil
+		result := combineResult(append(results, &res))
+		CheckWithM3DB(ctx, st, fetchOptions, q, cache, result)
+		return *result, nil
 	}
 	res, err := st.FetchProm(ctx, q, fetchOptions)
 	cache.cacheMetrics.CacheMetricsMiss(res)
 	return res, err
 }
+
+// Function to check cached results against M3DB results
+// Given a query, gets the result of that query from M3DB and compares it to cacheResult
+// NOTE: Especially for larger queries, cachedResult use M3DB for the most recent data
+// It is possible that the checking M3DB request that comes later has more data as more data loads in M3DB
+// As such, we do not record an error if we get a mismatch on data within a minute of the end query (but we still log it)
+// Note that requests to M3DB are not instant (though the queries to rulemanager are) M3coordinator has already parsed them into asking for
+// data from a specific range (instant queries would return nothing as nothing is in that range)
+func CheckWithM3DB(
+	ctx context.Context,
+	st storage.Storage,
+	fetchOptions *storage.FetchOptions,
+	q *storage.FetchQuery,
+	cache *RedisCache,
+	cacheResult *storage.PromResult,
+) {
+	if rand.Float32() < float32(cache.redisCacheSpec.CheckSampleRate) {
+		m3dbResult, err := st.FetchProm(ctx, q, fetchOptions)
+		if err == nil {
+			// Compare up to a bit ago to account for M3DB potentially not having been fully updated
+			equals := TimeseriesEqual(cacheResult.PromResult.Timeseries, m3dbResult.PromResult.Timeseries, q.End.Add(-CheckCutoffTime).UnixMilli(), cache.logger)
+			if !equals {
+				cache.logger.Info("Mismatch", zap.String("tags", q.TagMatchers.String()), zap.Int64("Start", q.Start.Unix()), zap.Int64("End", q.End.Unix()))
```

Comment:
Will update to error, the TimeseriesEqual function should log the difference, but I can also log expected/actual here though it will be quite a large log